### PR TITLE
feat: add AWS S3 storage plugin

### PR DESCRIPTION
**Pull request by MacAttak**
## Summary

Adds the native AWS S3 storage plugin package as an isolated provider-plugin slice.

This introduces `plugins/floe-storage-aws-s3` with:
- secret-free AWS S3 storage configuration
- `aws-s3` entry point for `floe.storage`
- `StorageDeploymentBinding` emission for native `s3` protocol
- support for environment, workload identity, and Kubernetes Secret credential reference modes
- non-secret PyIceberg, dbt, and Dagster runtime fragments
- package-local tests that can run both package-scoped and under root pytest collection

The public class names avoid the stale legacy `S3StoragePlugin` / `S3StorageConfig` names so the existing MinIO rename audits continue to enforce that the old generic `s3` plugin identity is gone.

## Validation

- `uv run pytest plugins/floe-storage-aws-s3/tests/unit/test_plugin.py tests/contract/test_storage_minio_reference_audit.py tests/contract/test_storage_minio_rename.py -q`
- `PYTHONPATH=plugins/floe-storage-aws-s3/src:packages/floe-core/src uv run ruff check plugins/floe-storage-aws-s3`
- `PYTHONPATH=plugins/floe-storage-aws-s3/src:packages/floe-core/src uv run mypy plugins/floe-storage-aws-s3/src/floe_storage_aws_s3`
- normal pre-push hooks passed: ruff, ruff format, mypy, dbt version requirements, bandit, uv-secure, unit tests, contract tests, no hardcoded sleep, requirement traceability, docs content validation, helm lint

### CODE REVIEW

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+from pydantic import ValidationError
+
+from floe_storage_aws_s3.config import AwsS3ObjectStoreConfig
+from floe_storage_aws_s3.plugin import AwsS3ObjectStorePlugin
+
+
+class TestPluginMetadata:
+    """Test plugin metadata and config schema."""
+
+    def test_plugin_metadata(self) -> None:
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py:32]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        plugin = AwsS3ObjectStorePlugin()
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py:34]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    @classmethod
+    def default_artifact_bucket(cls, data: Any) -> Any:
+        """Default artifact_bucket to bucket before frozen model creation."""
+        if isinstance(data, dict) and data.get("artifact_bucket") is None and "bucket" in data:
+            data = dict(data)
+            data["artifact_bucket"] = data["bucket"]
+        return data
+
+    @model_validator(mode="after")
+    def validate_credential_mode(self) -> AwsS3ObjectStoreConfig:
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        """Return non-secret PyIceberg catalog config for legacy callers."""
+        return self._pyiceberg_properties()
+
+    def _credential_binding(self) -> StorageCredentialBinding:
+        """Return secret-free credential references for the configured mode."""
+        config = self._require_config()
+        if config.credential_mode == "workload-identity":
+            if config.service_account_ref is None:
+                msg = "workload-identity credential_mode requires service_account_ref"
+                raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py:128]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
+"""Unit tests for the native AWS S3 storage plugin."""
+
+from __future__ import annotations
+
+from pathlib import Path
+from unittest.mock import patch
+
+import pytest
+import tomllib
+from pydantic import ValidationError
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py:10]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        payload = binding.model_dump_json()
+
+        assert binding.provider == "aws-s3"
+        assert binding.protocol == "s3"
+        assert binding.endpoint.internal_url == "https://s3.ap-southeast-2.amazonaws.com"
+        assert binding.endpoint.external_url == "https://s3.ap-southeast-2.amazonaws.com"
+        assert binding.endpoint.region == "ap-southeast-2"
+        assert binding.endpoint.warehouse_path == "s3://floe-provider-tests/warehouse/"
+        assert binding.endpoint.path_style_access is False
+        assert binding.warehouse is not None
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

File: plugins/floe-storage-aws-s3/tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+"""Shared pytest configuration for floe-storage-aws-s3 tests."""
+
+from __future__ import annotations
+
+import sys
+from pathlib import Path
+
+PACKAGE_SRC = Path(__file__).resolve().parents[1] / "src"
+if str(PACKAGE_SRC) not in sys.path:
+    sys.path.insert(0, str(PACKAGE_SRC))
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/conftest.py:13]: Fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1.

**Comment by MacAttak** [comment]
All seven review findings from claude issue comment 4429013519 were fixed in 1bf2af73c330c71d26015251b7bd76f8575e7bd1: requirement markers, test docstrings, Self validator return, redundant credential guards, test constants, warehouse assertion cleanup, and conftest rationale. The Python 3.10 CI failure was also fixed with a tomli fallback for tomllib.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        """Return secret-free credential references for the configured mode."""
+        config = self._require_config()
+        if config.credential_mode == "workload-identity":
+            assert config.service_account_ref is not None
+            return StorageCredentialBinding(
+                mode="workload-identity",
+                service_account_ref=config.service_account_ref,
+            )
+        if config.credential_mode == "kubernetes-secret":
+            assert config.credential_secret_name is not None
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    def _require_config(self) -> AwsS3ObjectStoreConfig:
+        """Return config or raise a structured plugin configuration error."""
+        if self._config is None:
+            from floe_core.plugin_errors import PluginConfigurationError
+
+            raise PluginConfigurationError(
+                "aws-s3",
+                [{"field": "_config", "message": "Plugin 'aws-s3' not configured"}],
+            )
+        return cast(AwsS3ObjectStoreConfig, self._config)
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+
+    @pytest.fixture()
+    def configured_plugin(self) -> AwsS3ObjectStorePlugin:
+        config = AwsS3ObjectStoreConfig(
+            bucket=TEST_BUCKET,
+            region=TEST_REGION,
+            credential_mode="workload-identity",
+            service_account_ref=TEST_SERVICE_ACCOUNT,
+        )
+        return AwsS3ObjectStorePlugin(config=config)
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py:351]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/tests/unit/test_plugin.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    @pytest.mark.requirement("STORAGE-AWS-S3-005")
+    def test_environment_config_rejects_secret_fields(self) -> None:
+        """Environment mode rejects Kubernetes Secret fields to keep sources exclusive."""
+        with pytest.raises(ValidationError, match="only accepts environment variable names"):
+            AwsS3ObjectStoreConfig(
+                bucket=TEST_BUCKET,
+                region=TEST_REGION,
+                credential_mode="environment",
+                credential_secret_name="aws-creds",  # pragma: allowlist secret
+            )
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/unit/test_plugin.py:80]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+AwsS3ObjectStoreCredentialMode = Literal["environment", "workload-identity", "kubernetes-secret"]
+CreatePolicy = Literal["must-exist", "never-create"]
+
+
+def _normalize_prefix(value: str) -> str:
+    """Normalize an S3 prefix to either empty or slash-terminated form."""
+    prefix = value.strip("/")
+    if not prefix:
+        return ""
+    return f"{prefix}/"
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py:19]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        if config.endpoint_override is not None:
+            dbt_profile_fragment["s3_endpoint"] = config.endpoint_override
+            dagster_resources["endpoint_url"] = config.endpoint_override
+
+        dbt_env_refs = (
+            {
+                "s3_access_key_id": AWS_ACCESS_KEY_ENV,
+                "s3_secret_access_key": AWS_SECRET_KEY_ENV,
+                "s3_session_token": AWS_SESSION_TOKEN_ENV,
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py]: Fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f.

File: plugins/floe-storage-aws-s3/tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+import sys
+from pathlib import Path
+
+# The package is intentionally not registered in the root workspace until the
+# provider integration branch. Package-local pytest already uses pythonpath,
+# but root-level CI collection needs this explicit source path.
+PACKAGE_SRC = Path(__file__).resolve().parents[1] / "src"
+if str(PACKAGE_SRC) not in sys.path:
+    sys.path.insert(0, str(PACKAGE_SRC))
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/tests/conftest.py:13]: Acknowledged. No code change required for this informational note; the temporary root-level CI path rationale remains documented in conftest until workspace registration lands.

**Comment by MacAttak** [comment]
Follow-up review summary 4429525346 fixed in 2a548ef6d050749db89c3a5bb20e0fdca9dfd82f: production asserts replaced with explicit guards, cross-mode rejection tests added, prefix edge cases covered, redundant config cast removed via runtime type guard, fixture docstring added, and dbt no longer requires AWS_SESSION_TOKEN for non-STS environment credentials.

**Comment by MacAttak** [comment]
Security review comment 4429499781 reviewed. The low-priority bucket/service-account format suggestions are not applicable for this slice: the plugin keeps only secret-free references, bucket existence/naming is externally owned by AWS/IaC, and there is no repo-wide service-account naming standard to enforce yet.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+        artifact_uri = _s3_uri(artifact_bucket, config.artifact_prefix)
+        credential_binding = self._credential_binding()
+        identity_modes = (
+            ["aws-irsa", "aws-pod-identity"]
+            if config.credential_mode == "workload-identity"
+            else []
+        )
+        runtime_env_refs = (
+            dict(credential_binding.env_refs) if config.credential_mode == "environment" else {}
+        )
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py:178]: Fixed in 61946f50df8efb71d52caba4e4affe7e1371e935.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+    bucket: str = Field(..., min_length=1, description="Warehouse bucket name.")
+    warehouse_prefix: str = Field(
+        default="warehouse/",
+        description="Warehouse prefix within the bucket.",
+    )
+    artifact_bucket: str | None = Field(
+        default=None,
+        description="Artifact bucket. Defaults to bucket when unset.",
+    )
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/config.py:41]: Fixed in 61946f50df8efb71d52caba4e4affe7e1371e935.

File: plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+        config = self._require_config()
+        return config.endpoint_override or f"https://s3.{config.region}.amazonaws.com"
+
+    def _pyiceberg_properties(self) -> dict[str, str]:
+        """Return non-secret PyIceberg S3 properties."""
+        config = self._require_config()
+        properties = {"s3.region": config.region}
+        if config.endpoint_override is not None:
+            properties["s3.endpoint"] = config.endpoint_override
+            properties["s3.path-style-access"] = str(config.path_style_access).lower()
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply plugins/floe-storage-aws-s3/src/floe_storage_aws_s3/plugin.py:109]: Fixed in 7e08ff30108d3c6673a7877c92d3d4edf964cadf.
